// mzf-schedule/dist/mzf-schedule.mjs
import { defineComponent as z, computed as R, ref as s, watch as Y, openBlock as d, createElementBlock as c, normalizeStyle as F, createElementVNode as o, toDisplayString as v, withDirectives as P, vShow as q, Fragment as S, renderList as M, withModifiers as G, createCommentVNode as H, pushScopeId as J, popScopeId as K } from "vue";
const $ = (r) => (J("data-v-33dc60eb"), r = r(), K(), r), Q = {
  class: "justify-between",
  "h-40px": "",
  flex: "",
  "items-center": "",
  "px-20px": "",
  "border-b": "1px dark:gray/10"
}, U = {
  flex: "~ gap3",
  text: "12px"
}, W = {
  flex: "~ gap2",
  "items-center": ""
}, Z = /* @__PURE__ */ $(() => /* @__PURE__ */ o("div", {
  "h-6px": "",
  "w-20px": "",
  "rounded-2px": "",
  "bg-blue-500": ""
}, null, -1)), ee = {
  flex: "~ gap2",
  "items-center": ""
}, te = /* @__PURE__ */ $(() => /* @__PURE__ */ o("div", {
  "h-6px": "",
  "w-20px": "",
  border: "0.5px dark:gray/50",
  "rounded-2px": ""
}, null, -1)), le = /* @__PURE__ */ $(() => /* @__PURE__ */ o("div", {
  flex: "~ gap2",
  "items-center": ""
}, [
  /* @__PURE__ */ o("div", { text: "gray" }, "可拖动鼠标选择时间段")
], -1)), ae = {
  "h-full": "",
  "w-full": "",
  "border-spacing-0": "",
  "table-fixed": ""
}, oe = { text: "14px" }, re = /* @__PURE__ */ $(() => /* @__PURE__ */ o("th", {
  rowspan: "2",
  "w-80px": "",
  "border-r": "1px dark:gray/10",
  "border-b": "1px dark:gray/10",
  "font-500": ""
}, " 星期/时间 ", -1)), ne = {
  colspan: "24",
  "border-r": "1px dark:gray/10",
  "border-b": "1px dark:gray/10"
}, ue = {
  colspan: "24",
  "border-b": "1px dark:gray/10"
}, se = ["border-r"], de = {
  "border-r": "1px dark:gray/10",
  "py-4px": "",
  text: "center 14px"
}, ce = ["border-r", "bg", "onMousedown", "onMousemove", "onClick"], ve = { key: 0 }, ie = {
  colspan: "49",
  "p-20px": "",
  text: "14px"
}, fe = {
  key: 0,
  text: "center"
}, pe = { key: 1 }, xe = { "pt-20px": "" }, he = { "min-w-50px": "" }, _e = /* @__PURE__ */ z({
  name: "VueSchedule",
  __name: "schedule",
  props: {
    value: null,
    width: { default: "800px" },
    selectedText: { default: "已选" },
    unselectedText: { default: "未选" },
    morningText: { default: "00:00 - 12:00" },
    afternoonText: { default: "12:00 - 24:00" },
    weeksText: { default: () => ["星期一", "星期二", "星期三", "星期四", "星期五", "星期六", "星期日"] },
    placeholder: { default: "可拖动鼠标选择时间段" },
    showResult: { type: Boolean, default: !1 }
  },
  emits: ["update:value", "formatTime"],
  setup(r, { emit: g }) {
    const i = r, f = R({
      get() {
        return i.value;
      },
      set(e) {
        g("update:value", e);
      }
    }), E = s(), D = s(!1), b = s(), y = s(), k = s(), m = s(), p = s(0), x = s(0), h = s(0), _ = s(0), u = s(Array.from(
      { length: 7 },
      (e, t) => Array.from({ length: 48 }, (l, a) => ({
        selected: !1,
        x: a,
        y: t
      }))
    ));
    Y(() => i.value, () => {
      if (f.value.length !== 336) {
        console.warn("value类型错误，value必须是0，1组成的336位字符串");
        return;
      }
      for (let e = 0; e < f.value.length; e++)
        f.value[e] === "0" ? u.value.flat()[e].selected = !1 : f.value[e] === "1" && (u.value.flat()[e].selected = !0);
    });
    const C = R(() => {
      const e = {}, t = i.weeksText;
      return u.value.forEach((l, a) => {
        l.forEach((n, T) => {
          n.selected && (e[t[a]] || (e[t[a]] = []), e[t[a]].push(T));
        }), e[t[a]] && (e[t[a]] = N(e[t[a]]));
      }), e;
    });
    Y(() => u.value, () => {
      g("update:value", j()), g("formatTime", C.value);
    }, { deep: !0 });
    function j() {
      let e = "";
      return u.value.flat().forEach((t) => {
        t.selected ? e += "1" : e += "0";
      }), e;
    }
    function L(e, t) {
      D.value = !0, b.value = e, y.value = t, document.addEventListener("mouseup", B);
    }
    function A(e, t) {
      D.value && (k.value = e, m.value = t);
    }
    function B() {
      D.value = !1;
      let e = !1;
      if (h.value === _.value && p.value === x.value) {
        V();
        return;
      }
      for (let t = h.value; t <= _.value; t++)
        for (let l = p.value; l <= x.value; l++)
          if (!u.value[t][l].selected) {
            e = !0;
            break;
          }
      if (e)
        for (let t = h.value; t <= _.value; t++)
          for (let l = p.value; l <= x.value; l++)
            u.value[t][l].selected = !0;
      else
        for (let t = h.value; t <= _.value; t++)
          for (let l = p.value; l <= x.value; l++)
            u.value[t][l].selected = !1;
      V();
    }
    function I(e, t) {
      return p.value = Math.min(b.value, k.value), x.value = Math.max(b.value, k.value), h.value = Math.min(y.value, m.value), _.value = Math.max(y.value, m.value), e >= p.value && e <= x.value && t >= h.value && t <= _.value;
    }
    function w(e) {
      const t = Math.floor(e / 2).toString().padStart(2, "0"), l = e % 2 === 0 ? "00" : "30";
      return `${t}:${l}`;
    }
    function N(e) {
      if (!e.length)
        return [];
      e.sort((a, n) => a - n);
      const t = [];
      let l = e[0];
      for (let a = 1; a < e.length; a++)
        if (e[a] - e[a - 1] !== 1) {
          const n = e[a - 1];
          t.push(`${w(l)} - ${w(n + 1)}`), l = e[a];
        }
      return t.push(`${w(l)} - ${w(e[e.length - 1] + 1)}`), t;
    }
    function V() {
      b.value = void 0, y.value = void 0, k.value = void 0, m.value = void 0, document.removeEventListener("mouseup", B);
    }
    function O() {
      u.value.flat().forEach((e) => {
        e.selected = !1;
      });
    }
    const X = R(
      () => !u.value.flat().some((e) => e.selected === !0)
    );
    return (e, t) => (d(), c("div", {
      style: F(`width:${r.width}`),
      "min-w-500px": "",
      bg: "dark:gray/10",
      "select-none": "",
      border: "1px dark:gray/5",
      "rounded-4px": ""
    }, [
      o("div", Q, [
        o("div", U, [
          o("div", W, [
            Z,
            o("div", null, v(r.selectedText), 1)
          ]),
          o("div", ee, [
            te,
            o("div", null, v(r.unselectedText), 1)
          ]),
          le
        ]),
        P(o("div", {
          text: "12px blue-500 hover:blue-500/80",
          "cursor-pointer": "",
          onClick: O
        }, " 清空 ", 512), [
          [q, !X.value]
        ])
      ]),
      o("div", null, [
        o("table", ae, [
          o("thead", null, [
            o("tr", oe, [
              re,
              o("th", ne, v(r.morningText), 1),
              o("th", ue, v(r.afternoonText), 1)
            ]),
            o("tr", null, [
              (d(), c(S, null, M(24, (l) => o("th", {
                key: l - 1,
                "border-r": l === 24 ? "none" : "1px dark:gray/10",
                colspan: "2",
                text: "13px center",
                "min-w-12px": "",
                "border-b": "1px dark:gray/10"
              }, v(l - 1), 9, se)), 64))
            ])
          ]),
          o("tbody", {
            ref_key: "tbodyRef",
            ref: E
          }, [
            (d(!0), c(S, null, M(u.value, (l, a) => (d(), c("tr", {
              key: a,
              "border-b": "1px dark:gray/10"
            }, [
              o("td", de, v(r.weeksText[a]), 1),
              (d(!0), c(S, null, M(u.value[a], (n) => (d(), c("td", {
                key: n.x,
                "border-r": n.x === 47 ? "none" : "1px dark:gray/10",
                colspan: "1",
                "min-w-12px": "",
                transition: "all 0.1s ease",
                bg: I(n.x, a) && n.selected ? "blue-500/50 " : I(n.x, a) ? "blue-500/10" : n.selected ? "blue dark:blue-500/60" : "hover:gray-500/10",
                onMousedown: (T) => L(n.x, a),
                onMousemove: (T) => A(n.x, a),
                onClick: G((T) => n.selected = !n.selected, ["prevent"])
              }, null, 40, ce))), 128))
            ]))), 128)),
            r.showResult ? (d(), c("tr", ve, [
              o("td", ie, [
                X.value ? (d(), c("div", fe, v(r.placeholder), 1)) : (d(), c("div", pe, [
                  o("div", xe, [
                    (d(!0), c(S, null, M(Object.keys(C.value), (l, a) => (d(), c("div", {
                      key: a,
                      flex: "~ gap5"
                    }, [
                      o("div", he, v(l), 1),
                      o("div", null, v(C.value[l].join("、")), 1)
                    ]))), 128))
                  ])
                ]))
              ])
            ])) : H("", !0)
          ], 512)
        ])
      ])
    ], 4));
  }
});
const ge = (r, g) => {
  const i = r.__vccOpts || r;
  for (const [f, E] of g)
    i[f] = E;
  return i;
}, ye = /* @__PURE__ */ ge(_e, [["__scopeId", "data-v-33dc60eb"]]);
export {
  ye as default
};
